release/src/router/plc-utils/nda/int6kp.1
.TH int6kp 1 "Mar 2014" "plc-utils-2.1.6" "Qualcomm Atheros Powerline Toolkit"

.SH NAME
int6kp - Production Device Manager 

.SH SYNOPSIS
.BR int6kp
.RI [ options ] 
.RI [ device ] 
.RI [ device ] 
[...]

.SH DESCRIPTION
The Production Device Manager performs basic device operations using the raw Ethernet protocol described in the Qualcomm Atheros HomePlug AV Firmware Technical Reference Manual.
It can be used to initialize devices, interrogate and control them and upgrade them when on-board NVRAM is present.

.PP
This program is part of the Qualcomm Atheros Powerline Toolkit.
It supports the legacy INT6000, INT6300, INT6400 and similar chipsets.
See program \fBamptool\fR to support INT7400, QCA7450 and similar chipsets.
See program \fBplctool\fR to support QCA6410, QCA7000, AR7420 and similar chipsets.

.PP
This program is the \fBproduction version\fR of \fBplc\fR and shall not be distributed without the explicit permission of Qualcomm Atheros, Ocala FL USA.

.SH OPTIONS

.TP
-\fBa\fR
Read device attributes using VS_OP_ATTRIBUTES.
Attributes are short strings and integers that describe device hardware and firmware.
They are concatenated to form the output that is similar to option -\fBr\fR but derived differently.

.TP
-\fBB \fIbutton\fR
Press the simple connect pushbutton using VS_PB_ENC.
The \fBaction\fR can be specified by number 1, 2, 3 or 4 or by symbol "\fBjoin\fR", "\fBleave\fR", "\fBstatus\fR" or "\fBreset\fR", respectively.
Use 1 on both devices that are expected to join.
Use 2 only on the device that is expected to leave the network.

.TP
-\fBC \fImodule\fR
Write (commit) downloaded modules to NVRAM using VS_MOD_NVM.
The \fBmodule\fR can be spedified by number 1=NVM, 2=PIB and 3=BOTH or string "nvm", "pib" or "both".
Module 3 is equivalent to option -\fBF\fR which writes the NVM and PIB together.
You cannot force flash NVRAM using this option.
Use option -\fBFF\fR to force flash.

.TP
-\fBd\fR \fIfilename\fR
Read Watchdog Report from the device and write it to the named file in binary format using VS_WD_RPT.
The report file can be sent to Qualcomm Atheros for technical analysis.
No assumptions are made based on filename and no filename convetions are enforced; however, you should use a \fB.log\fR file extension to indicate binary format.

.TP
-\fBD\fR \fIxx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx\fR
Define the 16 octet Device Access Key (DAK) in hex format.
The DAK is used by options -\fBJ\fR, -\fBC\fR and -\fBF\fR.
It may also be specified as "key1" or "key2" as explained in the \fBKEYS\fR section.
You may omit this option on the command line when the device DAK happens to be the same as the program default of "key1".

.TP
.RB - e
Redirects stderr messages to stdout.
By convention status and error messages are printed on stderr while primary program output is printed on stdout.
This option prints all output on stdout in cases where this is desired.

.TP
.RB - E
Erase NVRAM on the local device and reboot using VS_PTS_NVM.
This option is similar to option -\fBF\fR or -\fBFF\fR but writes all one's to NVRAM instead of downloaded firmware or paratemeters. Any firmware or parameters that have been downloaded pending flash will be lost when the device reboots. You must also specify the device \fBDAK\fR using option -\fBD\fR or the erase operation will fail without indication. This operation can only be performed on the local device. Attempts to erase a remote device are ignored. Some firmware versions do not permit the operation and those that do permit it will prevent it on INT6000 devices.
.TP
.RB - f
Read device NVRAM parameters using VS_GET_NVM. An error will be reported when no NVRAM is present. This option can be used to determine if NVRAM is large enough to store the runtime firmware and parameter block.
.TP
.RB - F [ F ]
Write previously downloaded MAC and PIB to NVRAM using VS_PTS_NVM. Adding a second \fBF\fR here or another -\fBF\fR anywhere on the command line will \fBforce-flash\fR a blank or corrupted NVRAM. Firmware loaded from NVRAM will treat force-flash as an error. This option will create new factory settings and erase any existing user settings. Subsequent use creates and updates operational settings that can be erased using a factory reset. This option is executed after all others on the command line, except for the -\fBR\fR option.
Unlike \fBplc\fR, this option will alter the factory default settings to match those of the PIB being written to NVRAM. Flashing with \fBplc\fR does not alter factory settings. 
.TP
.RB - g
Read multicast group information discovered while IGMP snooping using VS_MULTICAST_INFO.
.TP
-\fBi\fR \fIinterface\fR
Select the host Ethernet interface. All requests are sent via this host interface and only reponses received via this host interface are recognized. The default interface is \fBeth1\fR because most people use \fBeth0\fR as their principle network connection; however, if environment string "PLC" is defined then it takes precedence over the default interface. This option then takes precedence over either default.
.TP
.RB - I
Read the device PIB header using VS_RD_MOD and print the firmware major revision number, PIB minor revision number, Device Access Key (DAK), Network Membership Key (NMK), MAC address and other information on stdout.
.TP
.RB - J
Set the Network Membership Key (NMK) on a remote device using VS_SET_KEY. This option is similar to option -\fBK\fR but requires the remote device DAK in addition to NMK and MAC address. The NMK value is defined using option -\fBK\fR unless you want to use the default value. The remote DAK is defined using option -\fBD\fR unless you want to use the default value. Programming remote device keys is complicated. It is often easier to connect the device directly to the host and use the -\fBK\fR option.
.TP
-\fBK\fR \fIxx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx\fR
Define the Network Membership Key (NMK) value used by options -\fBM\fR or -\fBJ\fR. The symbolic names "key1" and "key2" are recognized as described in the \fBKEY\fR section.
.TP
-\fBl\fR \fIcount\fR
Define the number of times that the command will be repeated for each device specified. Normally, you will repeat operations on one device only.
.TP
.RB - m
Read network membership information using VS_NW_INFO. This can be used to determine network configuration. 
.TP
.RB - M
Set the Network Membership Key (NMK) on the local device using VS_SET_KEY. The NMK value is specified using the -\fBK\fR option unless you want to use the default value.  
.TP 
-\fBn\fR \fIfilename\fR
Read the firmware image chain from device flash memory and write it to the named \fB.nvm\fR file using multiple VS_RD_MOD messages. No assumptions are made based on filename and no filename conventions are enforced. This option is performed before option -\fBN\fR when both are specified.
.TP 
-\fBN\fR \fIfilename\fR
Read the named \fB.nvm\fR file and write it to the device using multiple VS_WR_MOD messages. No assumptions are made based on filename and no filename conventions are enforced; however, files having invalid .nvm format will be rejected. This option is executed after -\fBn\fR when both are specified.
.TP
-\fBp\fR \fIfilename\fR
Read user parameter block from device flash memory and write it to the named \fB.pib\fR file using multiple VS_RD_MOD messages. No assumptions are made based on filename and no filename convetions are enforced. This option is executed before option \f-BP\fR when both are specified.
.TP
-\fBP\fR \fIfilename\fR
Read the named \fB.pib\fR file and write it to the device using multiple VS_WR_MOD messages. No assumptions are made based on filename and no filename conventions are enforced; however, files having invalid .pib format will be rejected. This option is executed after \f-Bp\fR when both are specified.
.TP
.RB - q
Suppresses status messages on stderr. 
.TP
.RB - Q
Quick flash. The program will not wait for a device to reset or the firmware to restart after writing flash memory. This option is desirable with newer firmware that writes flash memory in the background. It has no effect unless used with option \fB-F\fR or \fB-C\fR.
.TP
.RB - r
Read device firmware and hardware revision using VS_SW_VER. Output is similar to option -\fBa\fR but is derived differently. Example output might look like "eth0 00:B0:52:00:00:01 INT6300A0 INT6000-MAC-2-0-2000-1018-20070611-SP1-B NOT-UPGRADEABLE".
.TP 
.RB - R
Reset the device using VS_RS_DEV. This option is executed after all others on the same command line.
.TP
.RB - s
Read device SDRAM configuration using VS_RD_CBLOCK.
.TP
-\fBS\fR \fIfilename\fR
Read the named \fB.cfg\fR file and write it to the device using VS_SET_SDRAM. No assumptions are make based on filename and no filename conventions are enforced; however, files having an invalid .cfg file format will be rejected. Do not confuse .cfg files with .config files. See \fBchkcfg\fR and \fBconfig2cfg\fR for more information on SDRAM configuration file formats. 
.TP
-\fBt \fItimeout\fR
Read timeout in milliseconds. Values range from 0 through UINT_MAX. This is the maximum time allowed for a response. The default is 50 milliseconds.
.TP
.RB - T
Restore factory defaults. This permanently erases PIB changes made since the device was last programmed with factory default values. The device will automatically reset and reboot.
.TP
.RB - v
Print additional information on stdout. In particular, this option dumps incoming and outgoing packets which can be saved as text files for reference.
.TP
-\fBw \fIseconds\fR
Defines the number of \fIseconds\fR to wait before repeating command line options. This option has no effect unless option -\fBl\fR is also specified with a non-zero value.
.TP
.RB - x
Cause the program to exit on the first error instead of continuing with remaining iterations, operations or devices. Normally, the program reports errors and moves on to the next operation, iteration or device depending on the command line.
.TP
.RB - ? ,-- help
Print program help summary on stdout. This option takes precedence over other options on the command line. 
.TP
.RB - ! ,-- version
Print program version information on stdout. This option takes precedence over other options on the command line. Use this option when sending screen dumps to Atheros Technical Support so that they know exactly which version of the Linux Toolkit you are using.
.SH ARGUMENTS
.TP
\fIdevice\fR
The MAC address of some powerline device. More than one address may be specified. If more than one address is specified then operations are performed on each device in turn. The default address is \fBlocal\fR. See \fBDEVICES\fR for information about symbolic device addresses.
.SH KEYS
Encryption keys are long, tedious to type and prone to error. For convenience, the following symbolic names may be used to specify either a Device Access Key or Network Membership Key. They may be entered in place of the full 16-octet key. The DAK and NMK will always be different for a given pass phrase.
.TP
.BR key1
This will translate to the HomePlug AV compliant DAK or NMK for pass phrase "HomePlugAV" depending on context.
.TP
.BR key2
This will translate to the HomePlug AV compliant DAK or NMK for pass phrase "HomePlugAV0123" depending on context.
.TP
.BR none
Symbolic name for "00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00".
.SH DEVICES
Powerline devices use Ethernet Media Access Control (MAC) addresses. A MAC address is a 48-bit value entered as 12 hexadecimal digits in upper, lower or mixed character case. Octets may be separated with colons for clarity. For example, "00b052000001", "00:b0:52:00:00:01" and "00b052:000001" are valid and equivalent.
.PP
The following MAC addresses are special and may be entered by name instead of number.
.TP
.BR all
Same as "broadcast".
.TP
.BR broadcast
A synonym for the Ethernet broadcast address, \fBFF:FF:FF:FF:FF:FF\fR. All devices, whether local, remote or foreign recognize messages sent to this address.  A remote device is any device at the far end of a powerline connection. A foreign device is any device not manufactured by Atheros.
.TP
.BR local
A synonym for the Qualcomm Atheros vendor specific Local Management Address (LMA), \fB00:B0:52:00:00:01\fR. All local Atheros devices recognize this address but remote and foreign devices do not. A remote device is any device at the far end of a powerline connection. A foreign device is any device not manufactured by Atheros.
.SH REFERENCES
See the Qualcomm Atheros HomePlug AV Firmware Technical Reference Manual for more information.
.SH DISCLAIMER
Atheros HomePlug AV Vendor Specific Management Message structure and content is proprietary to Qualcomm Atheros, Ocala FL USA. Consequently, public information is not available. Qualcomm Atheros reserves the right to modify message structure or content in future firmware releases without any obligation to notify or compensate users of this program.
.SH EXAMPLES
.PP
The following command performs 5 operations on one device. It uploads the firmware and PIB from the device and writes them to files \fBold.nvm\fR and \fBold.pib\fR, respectively. It reads files \fBnew.nvm\fR and \fBnew.pib\fR and downloads them as new firmware and PIB, respectively. It commits the downloaded firmware and PIB to NVRAM after validating the device DAK. Operations are executed in the order just described regardless of the order specified on the command line. If you want reading and writing to occur in a different order then you must use two or more commands to accomplish tasks in the order you want.
.PP
   # int6kp -n old.nvm -p old.pib -N new.nvm -P new.pib -D 00112233445566778899AABBCCDDEEFF -F 0123456789AB 
.PP
It is not neccessary to specify all operations on one command line. The three command lines below do essentially the same thing as the command line shown above. Notice that this example uses -\fBC 3\fR, instead of -\fBF\fR, as an alternate way to write both firmware and parameters to NVRAM. Specifying -\fBC 1\fR or -\fBC nvm\fR will write firmware only. Specifying, -\fBC 2\fR or -\fBC pib\fR will write parameters only. The value \fB3\fR is the logical OR of \fB1\fR and \fB2\fR and could be specified as -\fBC all\fR. The device DAK must appear on the same line as option \fB-C\fR because a valid DAK is required for the commit operation.
.PP
   # int6kp -N new.nvm 01:23:45:67:89:28
   # int6kp -P new.pib 01:23:45:67:89:28
   # int6kp -C 3 -D 00:11:22:33:44:55:66:77:88:99:BB:CC:DD:EE:FF 01:23:45:67:89:28 
.PP
The next example downloads file \fBnew.nvm\fR and file \fBnew.pib\fR and force flashes the \fBlocal\fR device. Force flashing only works with running firmware that has been downloaded and started by the Qualcomm Atheros Bootloader. See \fBplcfp\fR to download, start firmware and perform a force flash in one operation.
.PP
.PP
   # int6kp -N new.nvm -P new.pib -D 01:23:45:67:89:AB:CD:EF:01:23:45:67:89:AB:CD:EF -FF local 
.PP
The next two commands are equivalent. They set the NMK on the local device to \fBkey1\fR as descripted in the \fBKEYS\fR section. The first command resets the NMK on the local device with -\fBM\fR then specifies the NMK as \fBkey1\fR. The second command omits the key specification since \fBkey1\fR is the program default NMK. One could, of course, type the full encryption key.
.PP
   # int6kp -MK key1
   # int6kp -M
.PP
The next command erases the flash memory on the local device using \fBkey1\fR as the device DAK. Assuming that the device DAK matches, NVRAM will be erased. Some older versions of Atheros firmware do not support the erase operation.
.PP
   # int6kp -ED key1
.PP
The next command is similar to the previous example bu erases only flash memory sector \fB5\fR on the local device because option \fB-X\fR was specified. Again, some older versions of Atheros PLC firmware do not support this operation;
.PP
   # int6kp -X 5 -D key1
.SH SEE ALSO
.BR plc ( 1 ),
.BR plc ( 1 ),
.BR plcf ( 1 ),
.BR plcfp ( 1 ),
.BR plchost ( 1 ),
.BR plcid ( 1 ),
.BR plcrate ( 1 ),
.BR plcrule ( 1 ),
.BR plcstat ( 1 ),
.BR plcwait ( 1 )
.SH CREDITS
 Charles Maier <cmaier@qca.qualcomm.com>
 Nathaniel Houghton <nhoughto@qca.qualcomm.com>
'



release/src/router/plc-utils/nda/int6kp.c
/*====================================================================*
*
*   Copyright (c) 2013 Qualcomm Atheros, Inc.
*
*   All rights reserved.
*
*====================================================================*/

/*====================================================================*
 *
 *   int6kp.c -
 *
 *   Contributor(s):
 *      Charles Maier <cmaier@qca.qualcomm.com>
 *
 *--------------------------------------------------------------------*/

/*====================================================================*
 *   system header files;
 *--------------------------------------------------------------------*/

#include <unistd.h>
#include <stdlib.h>
#include <stdint.h>
#include <limits.h>

/*====================================================================*
 *   custom header files;
 *--------------------------------------------------------------------*/

#include "../tools/getoptv.h"
#include "../tools/putoptv.h"
#include "../tools/memory.h"
#include "../tools/number.h"
#include "../tools/symbol.h"
#include "../tools/types.h"
#include "../tools/flags.h"
#include "../tools/files.h"
#include "../tools/error.h"
#include "../key/HPAVKey.h"
#include "../key/keys.h"
#include "../ram/sdram.h"
#include "../pib/pib.h"
#include "../nvm/nvm.h"
#include "../plc/plc.h"
#include "../nda/nda.h"

/*====================================================================*
 *   custom source files;
 *--------------------------------------------------------------------*/

#ifndef MAKEFILE
#include "../plc/Attributes1.c"
#include "../plc/chipset.c"
#include "../plc/Confirm.c"
#include "../plc/Devices.c"
#include "../plc/Display.c"
#include "../plc/FactoryDefaults.c"
#include "../plc/Failure.c"
#include "../plc/Identity1.c"
#include "../plc/HostActionResponse.c"
#include "../plc/NetInfo1.c"
#include "../plc/NVRAMInfo.c"
#include "../plc/PushButton.c"
#include "../plc/ReadFirmware1.c"
#include "../plc/ReadParameters1.c"
#include "../plc/Request.c"
#include "../plc/ResetDevice.c"
#include "../plc/ReadMME.c"
#include "../plc/ReadMFG.c"
#include "../plc/SendMME.c"
#include "../plc/SetNMK.c"
#include "../plc/SDRAMInfo.c"
#include "../plc/StationRole.c"
#include "../plc/VersionInfo1.c"
#include "../plc/WaitForReset.c"
#include "../plc/WaitForRestart.c"
#include "../plc/WaitForStart.c"
#include "../plc/WriteCFG.c"
#include "../plc/WriteNVM.c"
#include "../plc/WritePIB.c"
#include "../plc/WatchdogReport.c"
#include "../plc/PLCSelect.c"
#endif

#ifndef MAKEFILE
#include "../nda/FlashPTS.c"
#include "../nda/EraseFlashMemory1.c"
#include "../nda/ReadFlashFirmware.c"
#include "../nda/ReadFlashParameters.c"
#endif

#ifndef MAKEFILE
#include "../tools/getoptv.c"
#include "../tools/putoptv.c"
#include "../tools/version.c"
#include "../tools/uintspec.c"
#include "../tools/basespec.c"
#include "../tools/hexdump.c"
#include "../tools/hexencode.c"
#include "../tools/hexdecode.c"
#include "../tools/hexstring.c"
#include "../tools/todigit.c"
#include "../tools/checkfilename.c"
#include "../tools/checksum32.c"
#include "../tools/fdchecksum32.c"
#include "../tools/error.c"
#include "../tools/strfbits.c"
#include "../tools/synonym.c"
#include "../tools/typename.c"
#endif

#ifndef MAKEFILE
#include "../ether/openchannel.c"
#include "../ether/closechannel.c"
#include "../ether/readpacket.c"
#include "../ether/sendpacket.c"
#include "../ether/channel.c"
#endif

#ifndef MAKEFILE
#include "../ram/nvram.c"
#include "../ram/sdramfile.c"
#include "../ram/sdrampeek.c"
#endif

#ifndef MAKEFILE
#include "../nvm/lightning_nvm_file.c"
#endif

#ifndef MAKEFILE
#include "../pib/lightning_pib_file.c"
#include "../pib/lightning_pib_peek.c"
#endif

#ifndef MAKEFILE
#include "../mme/MMECode.c"
#include "../mme/EthernetHeader.c"
#include "../mme/QualcommHeader.c"
#include "../mme/QualcommHeader1.c"
#include "../mme/UnwantedMessage.c"
#endif

#ifndef MAKEFILE
#include "../key/keys.c"
#endif

/*====================================================================*
 *   program constants;
 *--------------------------------------------------------------------*/

#define DEFAULT_LOOP 1
#define DEFAULT_WAIT 0

/*====================================================================*
 *   program variables;
 *--------------------------------------------------------------------*/

static struct _term_ const modules [] =

{
	{
		"nvm",
		"1"
	},
	{
		"pib",
		"2"
	},
	{
		"both",
		"3"
	}
};

static struct _term_ const buttons [] =

{
	{
		"join",
		"1"
	},
	{
		"leave",
		"2"
	},
	{
		"reset",
		"4"
	},
	{
		"status",
		"3"
	}
};

/*====================================================================*
 *
 *   void manager (struct plc *plc, signed loop, signed wait);
 *
 *   perform operations in logical order despite any order specfied
 *   on the command line; for example, read PIB before writing PIB;
 *
 *   operation order is controlled by the order of "if" statements
 *   shown here; the entire operation sequence can be repeated with
 *   an optional pause between each iteration;
 *
 *--------------------------------------------------------------------*/

static void manager (struct plc * plc, signed loop, signed wait)

{
	while (loop-- > 0)
	{
		if (_anyset (plc->flags, PLC_VERSION))
		{
			VersionInfo1 (plc);
		}
		if (_anyset (plc->flags, PLC_ATTRIBUTES))
		{
			Attributes1 (plc);
		}
		if (_anyset (plc->flags, PLC_WATCHDOG_REPORT))
		{
			WatchdogReport (plc);
		}
		if (_anyset (plc->flags, PLC_NVRAM_INFO))
		{
			NVRAMInfo (plc);
		}
		if (_anyset (plc->flags, PLC_SDRAM_CONFIG))
		{
			WriteCFG (plc);
		}
		if (_anyset (plc->flags, PLC_SDRAM_INFO))
		{
			SDRAMInfo (plc);
		}
		if (_anyset (plc->flags, PLC_READ_IDENTITY))
		{
			Identity1 (plc);
		}
		if (_anyset (plc->flags, PLC_NETWORK))
		{
			NetInfo1 (plc);
		}
		if (_anyset (plc->flags, PLC_READ_MAC))
		{
			ReadFlashFirmware (plc);
		}
		if (_anyset (plc->flags, PLC_WRITE_MAC))
		{
			WriteNVM (plc);
		}
		if (_anyset (plc->flags, PLC_READ_PIB))
		{
			ReadFlashParameters (plc);
		}
		if (_anyset (plc->flags, PLC_WRITE_PIB))
		{
			WritePIB (plc);
		}
		if (_anyset (plc->flags, PLC_HOST_ACTION))
		{
			HostActionResponse (plc);
		}
		if (_anyset (plc->flags, PLC_PUSH_BUTTON))
		{
			PushButton (plc);
		}
		if (_anyset (plc->flags, (PLC_SETLOCALKEY | PLC_SETREMOTEKEY)))
		{
			SetNMK (plc);
		}
		if (_anyset (plc->flags, PLC_FACTORY_DEFAULTS))
		{
			FactoryDefaults (plc);
		}
		if (_anyset (plc->flags, PLC_ERASE_DEVICE))
		{
			EraseFlashMemory1 (plc);
		}
		if (_anyset (plc->flags, PLC_FLASH_DEVICE))
		{
			FlashPTS (plc);
			WaitForRestart (plc);
		}
		if (_anyset (plc->flags, PLC_RESET_DEVICE))
		{
			ResetDevice (plc);
		}
		sleep (wait);
	}
	return;
}

/*====================================================================*
 *
 *   int main (int argc, char const * argv[]);
 *
 *   parse command line, populate plc structure and perform selected
 *   operations; show help summary if asked; see getoptv and putoptv
 *   to understand command line parsing and help summary display; see
 *   plc.h for the definition of struct plc;
 *
 *   the command line accepts multiple MAC addresses and the program
 *   performs the specified operations on each address, in turn; the
 *   address order is significant but the option order is not; the
 *   default address is a local broadcast that causes all devices on
 *   the local H1 interface to respond but not those at the remote
 *   end of the powerline;
 *
 *   the default address is 00:B0:52:00:00:01; omitting the address
 *   will automatically address the local device; some options will
 *   cancel themselves if this makes no sense;
 *
 *   the default interface is eth1 because most people use eth0 as
 *   their principle network connection; you can specify another
 *   interface with -i or define environment string PLC to make
 *   that the default interface and save typing;
 *
 *--------------------------------------------------------------------*/

int main (int argc, char const * argv [])

{
	extern struct channel channel;
	extern struct _term_ const devices [];
	extern struct _term_ const daks [];
	extern struct _term_ const nmks [];
	static char const * optv [] =
	{
		"AaB:C:d:D:eEfFi:IJ:K:l:mMn:N:qQp:P:rRS:st:Tvw:x12",
		"device [device] [...]",
		"Qualcomm Atheros INT6x00 Powerline Device Manager",
		"a\tread device attributes using VS_OP_ATTRIBUTES",
		"B n\tperform pushbutton action (n) using MS_PB_ENC [1|2|3|4|'join'|'leave'|'status'|'reset']",
		"C n\tflash NVRAM with module (n) using VS_MOD_NVM [1|2|3|'nvm'|'pib'|'both']",
		"d f\tdump and clear watchdog report to file (f) using VS_WD_RPT",
		"D x\tset DAK to (x) for option -E, -F and -J [" DAK1 "]",
		"e\tredirect stderr messages to stdout",
		"E\terase NVRAM then reset using VS_PTS_NVM",
		"f\tread NVRAM Configuration using VS_GET_NVM",
		"F[F]\tflash [force] NVRAM with firmware and PIB using VS_PTS_NVM",

#if defined (WINPCAP) || defined (LIBPCAP)

		"i n\thost interface is (n) [" LITERAL (CHANNEL_ETHNUMBER) "]",

#else

		"i s\thost interface is (s) [" LITERAL (CHANNEL_ETHDEVICE) "]",

#endif

		"I\tread device identity using VS_RD_MOD",
		"J\tchange Network Membership Key (NMK) on remote device using VS_SET_KEY (see -D and -K)",
		"K x\tset Network Membership Key (NMK) to (x) for options -J or -M [" NMK1 "]",
		"l n\tloop (n) times [" LITERAL (DEFAULT_LOOP) "]",
		"m\tread network membership information using VS_NW_INFO",
		"M\tchange Network Membership Key (NMK) on local device using VS_SET_KEY (see -K)",
		"n f\tread NVM from SDRAM to file (f) using VS_RD_MOD",
		"N f\twrite NVM file (f) to SDRAM using VS_WR_MOD",
		"p f\tread PIB from SDRAM to file (f) using VS_RD_MOD",
		"P f\twrite PIB file (f) to SDRAM using VS_WR_MOD",
		"q\tquiet mode",
		"Q\tquick flash",
		"r\tread hardware and firmware revision using VS_SW_VER",
		"R\treset device using VS_RS_DEV",
		"s\tread SDRAM Configuration using VS_RD_CBLOCK",
		"S f\tWrite an SDRAM Configuration file using VS_SET_SDRAM",
		"t n\tread timeout is (n) milliseconds [" LITERAL (CHANNEL_TIMEOUT) "]",
		"T\trestore factory defaults using VS_FAC_DEFAULTS",
		"v\tverbose mode",
		"w n\tpause (n) seconds [" LITERAL (DEFAULT_WAIT) "]",
		"x\texit on error",
		"1\tsuppress confirm MME when resetting device [option -R only]",
		"2\tboot from host after reset [option -R only]",
		(char const *) (0)
	};

#include "../plc/plc.c"

	signed loop = DEFAULT_LOOP;
	signed wait = DEFAULT_WAIT;
	signed c;
	if (getenv (PLCDEVICE))
	{

#if defined (WINPCAP) || defined (LIBPCAP)

		channel.ifindex = atoi (getenv (PLCDEVICE));

#else

		channel.ifname = strdup (getenv (PLCDEVICE));

#endif

	}
	optind = 1;
	while (~ (c = getoptv (argc, argv, optv)))
	{
		switch (c)
		{
		case 'a':
			_setbits (plc.flags, PLC_ATTRIBUTES);
			break;
		case 'B':
			_setbits (plc.flags, PLC_PUSH_BUTTON);
			plc.pushbutton = (unsigned) (uintspec (synonym (optarg, buttons, SIZEOF (buttons)), 0, UCHAR_MAX));
			break;
		case 'C':
			_setbits (plc.flags, PLC_FLASH_DEVICE);
			plc.module = (unsigned) (uintspec (synonym (optarg, modules, SIZEOF (modules)), 0, UCHAR_MAX));
			break;
		case 'd':
			_setbits (plc.flags, PLC_WATCHDOG_REPORT);
			if (! checkfilename (optarg))
			{
				error (1, EINVAL, "%s", optarg);
			}
			if ((plc.rpt.file = open (plc.rpt.name = optarg, O_BINARY | O_CREAT | O_RDWR | O_TRUNC, FILE_FILEMODE)) == -1)
			{
				error (1, errno, "%s", optarg);
			}

#ifndef WIN32

			chown (optarg, getuid (), getgid ());

#endif

			plc.readaction = 3;
			break;
		case 'D':
			if (! hexencode (plc.DAK, sizeof (plc.DAK), synonym (optarg, daks, SIZEOF (daks)))) 
			{ 
				error (1, errno, PLC_BAD_DAK, optarg);
			} 
			break;
		case 'e':
			dup2 (STDOUT_FILENO, STDERR_FILENO);
			break;
		case 'E':
			_clrbits (plc.flags, PLC_FLASH_DEVICE);
			_setbits (plc.flags, PLC_ERASE_DEVICE);
			plc.module = 0x0C;
			break;
		case 'f':
			_setbits (plc.flags, PLC_NVRAM_INFO);
			break;
		case 'F':
			_setbits (plc.module, (VS_MODULE_MAC | VS_MODULE_PIB));
			if (_anyset (plc.flags, PLC_FLASH_DEVICE))
			{
				_setbits (plc.module, VS_MODULE_FORCE);
			}
			_setbits (plc.flags, PLC_FLASH_DEVICE);
			break;
		case 'H':
			_setbits (plc.flags, PLC_HOST_ACTION);
			break;
		case 'I':
			_setbits (plc.flags, PLC_READ_IDENTITY);
			break;
		case 'i':

#if defined (WINPCAP) || defined (LIBPCAP)

			channel.ifindex = atoi (optarg);

#else

			channel.ifname = optarg;

#endif

			break;
		case 'J':
			if (! hexencode (plc.RDA, sizeof (plc.RDA), (char const *) (optarg)))
			{
				error (1, errno, PLC_BAD_MAC, optarg);
			}
			_setbits (plc.flags, PLC_SETREMOTEKEY);
			break;
		case 'K':
			if (! hexencode (plc.NMK, sizeof (plc.NMK), synonym (optarg, nmks, SIZEOF (nmks)))) 
			{ 
				error (1, errno, PLC_BAD_NMK, optarg);
			} 
			break;
		case 'M':
			_setbits (plc.flags, PLC_SETLOCALKEY);
			break;
		case 'l':
			loop = (unsigned) (uintspec (optarg, 0, UINT_MAX));
			break;
		case 'm':
			_setbits (plc.flags, PLC_NETWORK);
			break;
		case 'N':
			if (! checkfilename (optarg))
			{
				error (1, EINVAL, "%s", optarg);
			}
			if ((plc.NVM.file = open (plc.NVM.name = optarg, O_BINARY | O_RDONLY)) == -1)
			{
				error (1, errno, "%s", optarg);
			}
			if (lightning_nvm_file (& plc.NVM))
			{
				error (1, errno, "Bad thunderbolt image file: %s", plc.NVM.name);
			}
			_setbits (plc.flags, PLC_WRITE_MAC);
			break;
		case 'n':
			if (! checkfilename (optarg))
			{
				error (1, EINVAL, "%s", optarg);
			}
			if ((plc.nvm.file = open (plc.nvm.name = optarg, O_BINARY | O_CREAT | O_RDWR | O_TRUNC, FILE_FILEMODE)) == -1)
			{
				error (1, errno, "%s", optarg);
			}

#ifndef WIN32

			chown (optarg, getuid (), getgid ());

#endif

			_setbits (plc.flags, PLC_READ_MAC);
			break;
		case 'P':
			if (! checkfilename (optarg))
			{
				error (1, EINVAL, "%s", optarg);
			}
			if ((plc.PIB.file = open (plc.PIB.name = optarg, O_BINARY | O_RDONLY)) == -1)
			{
				error (1, errno, "%s", optarg);
			}
			if (lightning_pib_file (& plc.PIB))
			{
				error (1, errno, "Bad thunderbolt parameter file: %s", plc.PIB.name);
			}
			_setbits (plc.flags, PLC_WRITE_PIB);
			break;
		case 'p':
			if (! checkfilename (optarg))
			{
				error (1, EINVAL, "%s", optarg);
			}
			if ((plc.pib.file = open (plc.pib.name = optarg, O_BINARY | O_CREAT | O_RDWR | O_TRUNC, FILE_FILEMODE)) == -1)
			{
				error (1, errno, "%s", optarg);
			}

#ifndef WIN32

			chown (optarg, getuid (), getgid ());

#endif

			_setbits (plc.flags, PLC_READ_PIB);
			break;
		case 'Q':
			_setbits (plc.flags, PLC_QUICK_FLASH);
			break;
		case 'q':
			_setbits (channel.flags, CHANNEL_SILENCE);
			_setbits (plc.flags, PLC_SILENCE);
			break;
		case 'R':
			_setbits (plc.flags, PLC_RESET_DEVICE);
			break;
		case 'r':
			_setbits (plc.flags, PLC_VERSION);
			break;
		case 'S':
			if (! checkfilename (optarg))
			{
				error (1, EINVAL, "%s", optarg);
			}
			if ((plc.CFG.file = open (plc.CFG.name = optarg, O_BINARY | O_RDONLY)) == -1)
			{
				error (1, errno, "%s", optarg);
			}
			if (sdramfile (plc.CFG.file, optarg, plc.flags))
			{
				error (1, ECANCELED, "SDRAM config file %s is corrupt", optarg);
			}
			_setbits (plc.flags, PLC_SDRAM_CONFIG);
			break;
		case 's':
			_setbits (plc.flags, PLC_SDRAM_INFO);
			break;
		case 't':
			channel.timeout = (signed) (uintspec (optarg, 0, UINT_MAX));
			break;
		case 'T':
			_setbits (plc.flags, PLC_FACTORY_DEFAULTS);
			break;
		case 'v':
			_setbits (channel.flags, CHANNEL_VERBOSE);
			_setbits (plc.flags, PLC_VERBOSE);
			break;
		case 'w':
			wait = (unsigned) (uintspec (optarg, 0, 3600));
			break;
		case 'x':
			_setbits (plc.flags, PLC_BAILOUT);
			break;
		case '1':
			_setbits (plc.flag2, PLC_FLAG2_CFM_ACT);
			break;
		case '2':
			_setbits (plc.flag2, PLC_FLAG2_BT_SRC);
			break;
		default: 
			break;
		}
	}
	argc -= optind;
	argv += optind;
	if (argc != 1)
	{
		if (plc.nvm.file != -1)
		{
			error (1, ECANCELED, PLC_NODEVICE);
		}
		if (plc.pib.file != -1)
		{
			error (1, ECANCELED, PLC_NODEVICE);
		}
		if (plc.rpt.file != -1)
		{
			error (1, ECANCELED, PLC_NODEVICE);
		}
	}
	openchannel (& channel);
	if (! (plc.message = malloc (sizeof (* plc.message))))
	{
		error (1, errno, PLC_NOMEMORY);
	}
	if (! argc)
	{
		manager (& plc, loop, wait);
	}
	while ((argc) && (* argv))
	{
		if (! hexencode (channel.peer, sizeof (channel.peer), synonym (* argv, devices, SIZEOF (devices))))
		{
			error (1, errno, PLC_BAD_MAC, * argv);
		}
		manager (& plc, loop, wait);
		argc--;
		argv++;
	}
	free (plc.message);
	closechannel (& channel);
	exit (0);
}

